### PR TITLE
Zone11: different handling of root zone, TLD zones and zones in .arpa

### LINKED ISSUE #1256: SPF warning for TLD
With the MX check you explicitly check that a TLD should not have an MX.
But the SPF check, warns if a TLD does not have a SPF record.

The SPF standard explicitly tells clients not to check for SPF records on anything on the Public Suffix List. All TLDs are on the list.

### PR DESCRIPTION
**Pull request by marc-vanderwal**
## Purpose

This PR amends the specification for Zone11 so that different message tags of different severity levels are output if Zonemaster is run on the root zone, a TLD zone or a zone inside .arpa. It also cleans up some rough edges and migrates to MethodsV2.

## Context

See #1256.

## Changes

- Root zones, TLD zones and zones inside .arpa are tested differently: an INFO is output if no SPF or a null SPF is found, and a message tag of higher level is output if a non-null SPF is found.
- Use MethodsV2 instead of legacy Methods.
- Miscellaneous small rewordings

### Breaking message tag changes

This PR changes the arguments to the following existing message tags:

| Tag                              | Old arguments | New arguments |
|:---------------------------------|:--------------|:--------------|
| Z11_DIFFERENT_SPF_POLICIES_FOUND | ns_ip_list    | ns_list       |

This PR replaces the following message tags with new tags:

| Old tag                   | Old tag arguments | New tag                  | New tag arguments |
|:--------------------------|:------------------|:-------------------------|:------------------|
| Z11_SPF1_MULTIPLE_RECORDS | ns_ip_list        | Z11_SPF_MULTIPLE_RECORDS | ns_list           |
| Z11_SPF1_SYNTAX_ERROR     | ns_ip_list        | Z11_SPF_SYNTAX_ERROR     | ns_list           |
| Z11_SPF1_SYNTAX_OK        | ns_ip_list        | Z11_SPF_SYNTAX_OK        | ns_list           |

## How to test this PR

Review.

### CODE REVIEW

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 1]
```diff
@@ -33,6 +33,9 @@ It is assumed that *Child Zone* has been tested and reported by
 or name servers not giving a correct DNS response for an authoritative name
 server.
 
+The test is skipped if the zone being tested is the root zone, a [TLD] zone
+or a zone in the .ARPA tree.
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: I disagree, but the tags should be adjusted to those kind of zones.

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 2]
```diff
@@ -41,12 +44,13 @@ server.
...
 Z11_NO_SPF_FOUND                 | NOTICE  | domain             | No SPF policy was found for {domain}.
-Z11_SPF1_MULTIPLE_RECORDS        | ERROR   | ns_ip_list         | The following name servers returned more than one SPF policy. Name servers: {ns_ip_list}.
-Z11_SPF1_SYNTAX_ERROR            | ERROR   | domain, ns_ip_list | The SPF policy of {domain} has a syntax error. Policy retrieved from the following nameservers: {ns_ip_list}.
-Z11_SPF1_SYNTAX_OK               | INFO    | domain             | The SPF policy of {domain} has correct syntax.
+Z11_SPF_MULTIPLE_RECORDS         | ERROR   | ns_ip_list         | The following name servers returned more than one SPF policy. Name servers: {ns_ip_list}.
+Z11_SPF_SYNTAX_ERROR             | ERROR   | domain, ns_ip_list | The SPF policy of {domain} has a syntax error. Policy retrieved from the following nameservers: {ns_ip_list}.
+Z11_SPF_SYNTAX_OK                | INFO    | domain             | The SPF policy of {domain} has correct syntax.
+Z11_SKIPPING_NON_EMAIL_DOMAIN    | INFO    |                    | The test is skipped, because the domain is either the root name, a TLD or a zone in .ARPA.
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: See above.

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 3]
```diff
@@ -64,15 +68,18 @@ Response Defaults]. The handling of the DNS responses on the DNS queries follow,
 unless otherwise specified below, what is specified for [DNS Response] in the
 same specification.
 
-1. Create a [DNS Query] with query type TXT and query name *Child Zone* ("TXT
+1. If the *Child Zone* is the root zone ("."), a [TLD] or a zone in the .ARPA
+   [TLD], then output *[Z11_SKIPPING_NON_EMAIL_DOMAIN]* and terminate the test.
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: See above.

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 4]
```diff
@@ -41,12 +44,13 @@ server.
...
-Z11_INCONSISTENT_SPF_POLICIES    | WARNING |                    | One or more name servers do not publish the same SPF policy as the others.
 Z11_DIFFERENT_SPF_POLICIES_FOUND | NOTICE  | ns_ip_list         | The following name servers returned the same SPF policy, but other name servers returned a different policy. Name servers: {ns_ip_list}.
+Z11_INCONSISTENT_SPF_POLICIES    | WARNING |                    | One or more name servers do not publish the same SPF policy as the others.
 Z11_NO_SPF_FOUND                 | NOTICE  | domain             | No SPF policy was found for {domain}.
-Z11_SPF1_MULTIPLE_RECORDS        | ERROR   | ns_ip_list         | The following name servers returned more than one SPF policy. Name servers: {ns_ip_list}.
-Z11_SPF1_SYNTAX_ERROR            | ERROR   | domain, ns_ip_list | The SPF policy of {domain} has a syntax error. Policy retrieved from the following nameservers: {ns_ip_list}.
-Z11_SPF1_SYNTAX_OK               | INFO    | domain             | The SPF policy of {domain} has correct syntax.
+Z11_SPF_MULTIPLE_RECORDS         | ERROR   | ns_ip_list         | The following name servers returned more than one SPF policy. Name servers: {ns_ip_list}.
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: Here an other places we should include "ns_list" instead of "ns_ip_list" because it gives more information to the user.

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 5]
```diff
@@ -64,15 +68,18 @@ Response Defaults]. The handling of the DNS responses on the DNS queries follow,
...
+2. Create a [DNS Query] with query type TXT and query name *Child Zone* ("TXT
    query").
 
-2. Create an empty set of pairs of IP addresses and strings, "SPF-Policies".
+3. Create an empty set of pairs of IP addresses and strings, "SPF-Policies".
 
-3. Obtain the set of name server IP addresses using [Method4] and [Method5]
-   ("Name Server IP").
+4. Retrieve all name server IP addresses for *Child Zone* using methods
+   [Get-Del-NS-IPs] and [Get-Zone-NS-IPs] ("Name Server IP").
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: We should get name and IP of the name servers to be able to serve the messages with both.

**Review by matsduf — CHANGES_REQUESTED** [review]

**Comment by marc-vanderwal** [comment]
I have changed the approach and addressed your review comments. Please re-review.

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 6]
```diff
@@ -33,21 +33,29 @@ It is assumed that *Child Zone* has been tested and reported by
 or name servers not giving a correct DNS response for an authoritative name
 server.
 
+The root zone ("."), [TLD] zones and zones within .ARPA are treated
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: For me "under" sounds better.
```suggestion
The root zone ("."), [TLD] zones and zones under .ARPA are treated
```

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 7]
```diff
@@ -33,21 +33,29 @@ It is assumed that *Child Zone* has been tested and reported by
...
+differently because the corresponding domains are not expected to send e-mail.
+Additional discussion on the reason why these zones are unsuitable as e-mail
+domains is provided in the [Zone09 test specification].
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: The Zone09 test case specification excludes all zones under .arpa, but in Objective it only refer to in-addr.arpa and ip6.arpa zones. It should be clarified there that it is about any .arpa sub-zone. Maybe this PR could clarify in https://github.com/zonemaster/zonemaster/blob/master/docs/public/specifications/tests/Zone-TP/zone09.md?plain=1#L43-L47 that the exclusion is not only about reverse zones.

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 8]
```diff
@@ -33,21 +33,29 @@ It is assumed that *Child Zone* has been tested and reported by
...
-Z11_UNABLE_TO_CHECK_FOR_SPF      | ERROR   |                    | None of the zone’s name servers responded with an authoritative response to queries for SPF policies.
+Message Tag                      | Level   | Arguments       | Message ID for message tag
+:--------------------------------|:--------|:----------------|:--------------------------------------------
+Z11_DIFFERENT_SPF_POLICIES_FOUND | NOTICE  | ns_list         | The following name servers returned the same SPF policy, but other name servers returned a different policy. Name servers: {ns_list}.
+Z11_INCONSISTENT_SPF_POLICIES    | WARNING |                 | One or more name servers do not publish the same SPF policy as the others.
+Z11_NO_SPF_FOUND                 | NOTICE  | domain          | No SPF policy was found for {domain}.
+Z11_NO_SPF_NON_MAIL_DOMAIN       | INFO    | domain          | No SPF policy was found for {domain}, which is not an e-mail domain.
+Z11_NON_NULL_SPF_NON_MAIL_DOMAIN | NOTICE  | domain          | A non-null SPF policy was found on {domain}, although it is not an e-mail domain.
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: I think that "not an e-mail domain" is not precise enough. Any domain without MX could be said to be "not an e-mail domain".
```suggestion
Z11_NON_NULL_SPF_NON_MAIL_DOMAIN | NOTICE  | domain          | A non-null SPF policy was found on {domain}, although this type of domain (root, TLD or under .ARPA) is unexpected to be used for e-mail.
```

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 9]
```diff
@@ -33,21 +33,29 @@ It is assumed that *Child Zone* has been tested and reported by
...
+Z11_NULL_SPF_NON_MAIL_DOMAIN     | INFO    | domain          | A null SPF policy was found on {domain}, which is not an e-mail domain.
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]:
```suggestion
Z11_NULL_SPF_NON_MAIL_DOMAIN     | INFO    | domain          | A null SPF policy was found on {domain}, which is a type of domain (root, TLD or under .ARPA) which is unexpected to be used for e-mail.
```

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 10]
```diff
@@ -67,70 +75,86 @@ same specification.
...
+   "SPF-Policies".
 
-3. Obtain the set of name server IP addresses using [Method4] and [Method5]
-   ("Name Server IP").
+3. Retrieve all name server names and IP addresses for *Child Zone* using
+   methods [Get-Del-NS-Names-and-IPs] and [Get-Zone-NS-Names-and-IPs] ("Name
+   Servers").
 
-4. For each name server in *Name Server IP* do:
+4. For each name server in *Name Servers* do:
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: This PR has chosen a different approach than #1409 for DNSSEC05. In the implementation the name server IP what counts. In the messages name/IP should be included. I think we should find a wording that we consistently use in all updates where name/IP are expected in the messages.

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 11]
```diff
@@ -67,70 +75,86 @@ same specification.
...
    the same string ("SPF policy").
 
 10. If the *SPF Policy* does not [pass the syntax check][passing the syntax
-    check] for SPF version 1 records, then output *[Z11_SPF1_SYNTAX_ERROR]* and
-    terminate the test.
+    check] for SPF records, then output *[Z11_SPF_SYNTAX_ERROR]* and terminate
+    the test.
+
+11. If the *Child Zone* is the root zone ("."), a [TLD] or a zone within
+    .ARPA, then:
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]:
```suggestion
11. If the *Child Zone* is the root zone ("."), a [TLD] or a zone under
    .ARPA, then:
```

**Review by matsduf — DISMISSED** [review]

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 6 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: Sounds good indeed.

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 7 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: I’d rather create a seperate PR for clarifying that paragraph in Zone09. Meanwhile, I’ll just rewrite this paragraph here so that the text does not refer to Zone09 anymore.

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 8 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: Good suggestion.

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 9 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: Good suggestion, but I’ll slightly reword in order to avoid repeating “which is” twice in the same sentence.

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 10 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: Okay, I’ll reword as “For each name server IP address in *Name Server IP* do:”

Comment (matsduf) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: Well, there are some aspects of the approach in #1409. First there is the text in https://github.com/zonemaster/zonemaster/blob/e8b95a6cfd8d43d59577c92b94b179033c9c0aa5/docs/public/specifications/tests/DNSSEC-TP/dnssec05.md?plain=1#L115-L118 and then https://github.com/zonemaster/zonemaster/blob/e8b95a6cfd8d43d59577c92b94b179033c9c0aa5/docs/public/specifications/tests/DNSSEC-TP/dnssec05.md?plain=1#L175-L178

I am not sure if the approach in #1409 is the best one. In the implementation you probably want to go by IP address, and should that be reflected in the specification?

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 7 above)
Comment (matsduf) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: I think there is a point keeping the two specifications the same in this respect.

Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: It’s better to avoid cross-referencing specifications too much. Besides, the reason why we treat the root, TLD zones and .ARPA differently here is slightly different from Zone09: in Zone09, it’s because the resulting email addresses are unlikely to make sense as recipient addresses, but in Zone11, it’s because they are unlikely to make sense as sender addresses.

I took the liberty to make minor adjustments (other than what is strictly necessary) to the text instead of merely copy-pasting it from Zone09. What do you think? Maybe we can import my text back into Zone09 later?

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 10 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: Oh, I didn’t notice those bits in the DNSSEC05 proposal.

Yes, it is a recurring pattern that we iterate on *name server IPs* under the hood but that we report *name server name and IP pairs* in user-visible messages. And we need to be careful with the edge case where a zone has more than one name server name pointing to the same IP address.

So I think the approach makes sense. Maybe we should keep it somewhere in a template?

Comment (matsduf) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: It is not only the edge case when a zone has serveral names to the same IP address, it is also the case when the name in the delegation does not match the name in the zone, but the IP is the same. The latter is probably more common.

We have a template in https://github.com/zonemaster/zonemaster/blob/master/docs/internal/templates/specifications/tests/Template01.md which probably needs some updates.

**Review by matsduf — DISMISSED** [review]
Looks fine. Do you plan to create a scenario PR too?

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 12]
```diff
@@ -1,7 +1,7 @@
-# ZONE11: SPF policy validation
+# Zone11: SPF policy validation
 
 ## Test case identifier
-**ZONE11**
+**Zone11**
```
Comment (tgreenx) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: This should probably not be lowercased now. I'm not sure if we even want that here, but if we do, it might be best to do it for all specifications at once so we do not get any discrepancies.

**Review by tgreenx — COMMENTED** [review]
LGTM. Just one comment.

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 13]
```diff
@@ -39,15 +63,18 @@ server.
...
-Z11_NO_SPF_FOUND                 | NOTICE  | domain             | No SPF policy was found for {domain}.
-Z11_SPF1_MULTIPLE_RECORDS        | ERROR   | ns_ip_list         | The following name servers returned more than one SPF policy. Name servers: {ns_ip_list}.
-Z11_SPF1_SYNTAX_ERROR            | ERROR   | domain, ns_ip_list | The SPF policy of {domain} has a syntax error. Policy retrieved from the following nameservers: {ns_ip_list}.
-Z11_SPF1_SYNTAX_OK               | INFO    | domain             | The SPF policy of {domain} has correct syntax.
-Z11_UNABLE_TO_CHECK_FOR_SPF      | ERROR   |                    | None of the zone’s name servers responded with an authoritative response to queries for SPF policies.
+Message Tag                      | Level   | Arguments       | Message ID for message tag
+:--------------------------------|:--------|:----------------|:--------------------------------------------
+Z11_DIFFERENT_SPF_POLICIES_FOUND | NOTICE  | ns_list         | The following name servers returned the same SPF policy, but other name servers returned a different policy. Name servers: {ns_list}.
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: Z11_DIFFERENT_SPF_POLICIES_FOUND is always outputted together with Z11_INCONSISTENT_SPF_POLICIES. The message ID could the be shortened to `The following name servers returned the same SPF policy. Name servers: {ns_list}.`

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 12 above)
Comment (matsduf) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: As far as I can see it is always upper case, but we often refer to the test cases with mixed case (except DNSSEC). We should probably keep upper case as @tgreenx suggests.

Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: Okay, I’ll revert that change.

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 14]
```diff
@@ -67,70 +98,90 @@ same specification.
...
           string thus derived from the RDATA field to the *SPF-Policies* set.
 
    6. Go to the next name server.
 
 5. If the *SPF-Policies* set is empty, then output
    *[Z11_UNABLE_TO_CHECK_FOR_SPF]* and terminate the test.
 
-6. If all the pairs in the *SPF-Policies* set contain empty strings, then
-   output *[Z11_NO_SPF_FOUND]* and terminate the test.
+6. If all the pairs in the *SPF-Policies* set contain empty strings, then:
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: I think the following is easier to read.

```suggestion
6. If all the name server IPs in the *SPF-Policies* set contain empty strings (no "SPF policy"), then:
```

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 15]
```diff
@@ -67,70 +98,90 @@ same specification.
...
+   2. Else, output *[Z11_NO_SPF_FOUND]* and terminate the test.
 
-7. Compare the set of *SPF-Policies* retrieved from all name servers. If at
+7. For all messages outputted below, if an IP address in *Name Servers* is
+   connected to more than one name server name, then all names should be
+   included with the message tag.
+
+8. Compare the set of *SPF-Policies* retrieved from all name servers. If at
    least two different name servers have returned different sets of SPF
    policies, then:
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]:
```suggestion
8. Compare the set of *SPF-Policies* retrieved from all name servers (in the
   the *SPF-Policies* set). If at least two different name servers have returned 
   different sets of SPF policies, then:
```

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 16]
```diff
@@ -67,70 +98,90 @@ same specification.
...
 
-8. If the *SPF-Policies* set contains at least two entries with the same IP
-   address, then output *[Z11_SPF1_MULTIPLE_RECORDS]* with the list of
+9. If the *SPF-Policies* set contains at least two entries with the same IP
+   address, then output *[Z11_SPF_MULTIPLE_RECORDS]* with the list of
    nameservers that returned more than one SPF policy and terminate the test.
 
-9. The following steps assume that all pairs in the *SPF-Policies* set have
-   the same string ("SPF policy").
+10. The following steps assume that all pairs in the *SPF-Policies* set have
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]:
```suggestion
10. The following steps assume that all name server IPs in the *SPF-Policies* set have
```

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 17]
```diff
@@ -67,70 +98,90 @@ same specification.
...
+12. If the *Child Zone* is the root zone ("."), a [TLD] or a zone under
+    .ARPA, then:
 
-11. If no other message was outputted by this test case, then output
-    *[Z11_SPF1_SYNTAX_OK]*.
+   1. If the *SPF Policy* is a [Null SPF] policy, then output
+      *[Z11_NULL_SPF_NON_MAIL_DOMAIN]* and terminate the test.
+
+   2. If the *SPF Policy* is not a [Null SPF] policy, then output
+      *[Z11_NON_NULL_SPF_NON_MAIL_DOMAIN]* and terminate the test.
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: Not correctly indented in the Github rendering. I think a leading space is missing.

```suggestion
    1. If the *SPF Policy* is a [Null SPF] policy, then output
       *[Z11_NULL_SPF_NON_MAIL_DOMAIN]* and terminate the test.

    2. If the *SPF Policy* is not a [Null SPF] policy, then output
       *[Z11_NON_NULL_SPF_NON_MAIL_DOMAIN]* and terminate the test.
```

**Review by matsduf — CHANGES_REQUESTED** [review]

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 17 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: Indeed. This is fixed and I checked myself that it renders properly on GitHub.

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 16 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: Fixed.

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 15 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: Fixed.

File: docs/public/specifications/tests/Zone-TP/zone11.md (hunk 14 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Zone-TP/zone11.md]: Fixed.

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 18]
```diff
@@ -67,70 +98,91 @@ same specification.
...
+8. Compare the set of *SPF-Policies* retrieved from all name servers (in the
+   *SPF-Policies* set). If at least two different name servers have returned
+   different sets of SPF policies, then:
 
    1. Output *[Z11_INCONSISTENT_SPF_POLICIES]*.
    2. Group *SPF-Policies* by equal sets of SPF policies, such that a set of
-      SPF policies is mapped to the list of *Name Server IPs* that returned it.
+      SPF policies is mapped to the list of *Name Servers* that returned it.
    3. For each such group of name servers, output
       *[Z11_DIFFERENT_SPF_POLICIES_FOUND]*.
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: It should be stated that the list of name servers ("ns_list") is included with the tag. This applies to other places too.

File: docs/public/specifications/tests/Zone-TP/zone11.md [hunk 19]
```diff
@@ -67,70 +98,91 @@ same specification.
...
-6. If all the pairs in the *SPF-Policies* set contain empty strings, then
-   output *[Z11_NO_SPF_FOUND]* and terminate the test.
+6. If all the name server IPs in the *SPF-Policies* set contain empty strings
+   (no "SPF policy"), then:
+
+   1. If the *Child Zone* is the root zone ("."), a [TLD] or a zone under
+      .ARPA, then output *[Z11_NO_SPF_NON_MAIL_DOMAIN]* and terminate the
+      test.
+
+   2. Else, output *[Z11_NO_SPF_FOUND]* and terminate the test.
```
Comment (matsduf) [inline docs/public/specifications/tests/Zone-TP/zone11.md]: It should be stated that the domain is outputted with the tag.

**Review by matsduf — CHANGES_REQUESTED** [review]
The test case has "ns_list" arguments for some tags, and then there is a chance that the same IP address is used by multiple name server names. It should be stated explicitly that "ns_list" includes all such name/IP pairs.

**Comment by marc-vanderwal** [comment]
Step 7 in the test procedure already states:

> For all messages outputted below, if an IP address in *Name Servers* is connected to more than one name server name, then all names should be included with the message tag.

But I found a paragraph in the “Summary” section that was no longer true. I’ve updated it with a similar text. Does this make step 7 in the test procedure redundant?

**Review by matsduf — APPROVED** [review]

**Review by tgreenx — APPROVED** [review]
LGTM.

Implementation in https://github.com/zonemaster/zonemaster-engine/pull/1477.

**Review by tgreenx — APPROVED** [review]

**Comment by matsduf** [comment]
@marc-vanderwal, please merge.
